matrix-notifications.el

```diff
@@ -0,0 +1,58 @@
+
+(defvar matrix-client-notify-hook
+  '(matrix-client-notify)
+  "List of functions called for events.
+Each is called with the event-type and the event data.")
+
+(defvar matrix-client-notifications nil
+  "Alist of recent notifications mapping notification ID to related buffer.
+Automatically trimmed to last 20 notifications.")
+
+(defun matrix-client-notify (event-type data &rest rest)
+  "Notify for an event of EVENT-TYPE with DATA."
+  (let ((fn (intern-soft (concat "matrix-client-notify-" event-type))))
+    (when (functionp fn)
+      (apply #'funcall fn data rest))))
+
+;; MAYBE: Use a macro to define the handlers, because they have to
+;; define their arg lists in a certain way, and the macro would take
+;; care of that automatically.
+
+(cl-defun matrix-client-notify-m.room.message (data &key room &allow-other-keys)
+  "Show notification for m.room.message events.
+DATA should be the `data' variable from the
+`defmatrix-client-handler'.  ROOM should be the room object."
+  (pcase-let* (((map content sender event_id) data)
+               ((map body) content)
+               (display-name (matrix-client-displayname-from-user-id room sender))
+               (buffer (oref room :buffer))
+               (id (notifications-notify :title (format "<b>%s</b>" display-name)
+                                         :body body
+                                         :category "im.received"
+                                         :timeout 5000
+                                         :app-icon nil
+                                         :actions '("default" "Show")
+                                         :on-action #'matrix-client-notification-show)))
+    (map-put matrix-client-notifications id (a-list 'buffer buffer
+                                                    'event_id event_id))
+    ;; Trim the list
+    (setq matrix-client-notifications (-take 20 matrix-client-notifications))))
```

Comment:
The '20' should be configurable, but I'm fine with tackling that later on.

Reply:
Yeah, there are several things that will need to be defcustoms.  :)